### PR TITLE
TY: Fixed type inference for `Self {}` struct literal

### LINKED ISSUE #2361: Unresolved reference `None`
In this code:

```rust
struct Node<'a> {
    depth: u32,
    zero: Option<&'a Node<'a>>,
    one: Option<&'a Node<'a>>,
    value: Option<u32>,
}

impl<'a> Node<'a> {
    pub fn new() -> Self {
        Self {
            depth: 0,
            zero: None,
            one: None,
            value: None,
        }
    }
}

fn main() {
    let root = Node::new();
    println!("{}", root.value.is_some());
}
```

None in `new` function are marked as Unresolved, but `cargo check` show just warnings for never used fields.

unresolver reference

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #2361

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt [hunk 1]
```diff
@@ -128,3 +126,40 @@ fun resolvePath(path: RsPath, lookup: ImplLookup = ImplLookup.relativeTo(path)):
 private fun resolveAssocTypeBinding(trait: RsTraitItem, binding: RsAssocTypeBinding): RsTypeAlias? =
     collectResolveVariants(binding.referenceName) { processAssocTypeVariants(trait, it) }
         .singleOrNull() as? RsTypeAlias?
+
+/** Resolves a reference through type aliases */
+fun RsPathReference.throughResolve(): RsElement? =
+    advancedThroughResolve()?.element
+
+/** Resolves a reference through type aliases */
+fun RsPathReference.advancedThroughResolve(): BoundElement<RsElement>? {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt]: Maybe `deepResolve` is better?

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt [hunk 2]
```diff
@@ -128,3 +126,40 @@ fun resolvePath(path: RsPath, lookup: ImplLookup = ImplLookup.relativeTo(path)):
...
+
+    // Resolve potential type aliases
+    return if (boundElement != null && boundElement.element is RsTypeAlias) {
+        resolveThroughTupeAliases(boundElement)
+    } else {
+        boundElement
+    }
+}
+
+private fun resolveThroughTupeAliases(boundElement: BoundElement<RsElement>): BoundElement<RsElement>? {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt]: `Tupe` -> `Type`

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt [hunk 3]
```diff
@@ -128,3 +126,40 @@ fun resolvePath(path: RsPath, lookup: ImplLookup = ImplLookup.relativeTo(path)):
...
+    var base: BoundElement<RsElement> = boundElement
+    @Suppress("NAME_SHADOWING")
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt]: ^

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/ref/RsPathReferenceImpl.kt]: I hesitated a long time with the choice, but ok, will be `deepResolve` :)

**Comment by vlad20012** [comment]
bors r=undin

**Pull request merged**